src/algebra/ordered_sub.lean

```diff
@@ -0,0 +1,680 @@
+/-
+Copyright (c) 2021 Floris van Doorn. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Floris van Doorn
+-/
+import algebra.ordered_monoid
+/-!
+# Ordered Subtraction
+
+This file proves lemmas relating (truncated) subtraction with an order. We provide a class
+ `has_ordered_sub` stating that `a - b ≤ c ↔ a ≤ c + b`.
+
+The subtraction discussed here could both be normal subtraction in an additive group or truncated
+subtraction on a canonically ordered monoid (`ℕ`, `multiset`, `enat`, `ennreal`, ...)
+
+## Implementation details
+
+`has_ordered_sub` is a mixin type-class, so that we can use the results in this file even in cases
+where we don't have a `canonically_ordered_add_monoid` instance
+(even though that is our main focus). Conversely, this means we can use
+`canonically_ordered_add_monoid` without necessarily having to define a subtraction.
+
+The results in this file are ordered by the type-class assumption needed to prove it.
+This means that similar results might not be close to each other. Furthermore, we don't prove
+implications if a bi-implication can be proven under the same assumptions.
+
+Many results about this subtraction are primed, to not conflict with similar lemmas about ordered
+groups.
+
+We provide a second version of most results that require `[contravariant_class α α (+) (≤)]`. In the
+second version we replace this type-class assumption by explicit `add_le_cancellable` assumptions.
+
+TODO: maybe we should make a multiplicative version of this, so that we can replace some identical
+lemmas about subtraction/division in `ordered_[add_]comm_group` with these.
+
+-/
+
+variables {α : Type*}
+
+/-- `has_ordered_sub α` means that `α` has a subtraction characterized by `a - b ≤ c ↔ a ≤ c + b`.
+In other words, `a - b` is the least `c` such that `a ≤ b + c`.
+
+This is satisfied both by the subtraction in additive ordered groups and by truncated subtraction
+in canonically ordered monoids on many specific types.
+-/
+class has_ordered_sub (α : Type*) [preorder α] [has_add α] [has_sub α] :=
+(sub_le_iff_right : ∀ a b c : α, a - b ≤ c ↔ a ≤ c + b)
+
+section ordered_add_comm_monoid
+variables {a b c d : α} [partial_order α] [add_comm_monoid α] [has_sub α] [has_ordered_sub α]
+
+@[simp] lemma sub_le_iff_right : a - b ≤ c ↔ a ≤ c + b :=
```

Comment:
Could you explain how you selected this set of lemmas for the remainder of the file? Is it based on what we have already proven for ℕ + any intermediate results needed for them?

Reply:
It was mostly the stuff proven for `nat`;
Some of it was proven for another specific type like `ennreal`; 
Some of the lemmas were variations on already existing lemmas (for example: for `nat` and `iff` was proven, and `contravariant_class` is needed for one direction. That results in 3 lemmas: the direction that always holds without assumptions, the `iff` under an `add_le_cancellable` assumption and the `iff` under an `contravariant_class` assumption);
Some of the lemmas are analogues of subtraction lemmas from `ordered_group` (often with additional assumptions in this file).